src/doc/frames_sequence.cpp

```diff
@@ -0,0 +1,207 @@
+// Aseprite Document Library
+// Copyright (c) 2023 Igara Studio S.A.
+//
+// This file is released under the terms of the MIT license.
+// Read LICENSE.txt for more information.
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "doc/frames_sequence.h"
+
+#include "base/base.h"
+#include "base/debug.h"
+#include "base/serialization.h"
+
+#include <algorithm>
+#include <iostream>
+
+namespace doc {
+
+using namespace base::serialization;
+using namespace base::serialization::little_endian;
+
+FramesSequence::FramesSequence(const SelectedFrames& selectedFrames)
+{
+  // TODO: This might be optimized by copying each range directly, but
+  // for this I would need that SelectedFrames make this class its friend.
+  for(auto frame : selectedFrames) {
+    this->insert(frame);
+  }
+}
+
+std::size_t FramesSequence::size() const
+{
+  std::size_t size = 0;
+  for (auto& range : m_ranges)
+    size += ABS(range.toFrame - range.fromFrame) + 1;
+  return size;
+}
+
+void FramesSequence::clear()
+{
+  m_ranges.clear();
+}
+
+void FramesSequence::insert(frame_t frame)
+{
+  ASSERT(frame >= 0);
+
+  if (m_ranges.empty()) {
+    m_ranges.push_back(FrameRange(frame));
+    return;
+  }
+
+  auto& lastRange = m_ranges.back();
+  int rangeStep = (lastRange.fromFrame < lastRange.toFrame) -
+                  (lastRange.fromFrame > lastRange.toFrame);
+  int step = (lastRange.toFrame < frame) - (lastRange.toFrame > frame);
+  if (step &&
+      (rangeStep == step || !rangeStep) &&
+      frame-lastRange.toFrame == step)
+    lastRange.toFrame = frame;
+  else
+    m_ranges.push_back(FrameRange(frame));
+}
+
+void FramesSequence::insert(frame_t fromFrame, frame_t toFrame)
+{
+  int step = fromFrame <= toFrame ? 1 : -1;
```

Comment:
Please add `(...)` in these kind of operators.
```suggestion
  int step = (fromFrame <= toFrame ? 1 : -1);
```